File: GridView/AllowEndUsersAddItemsComboBoxEditor/AllowEndUsersAddItemsComboBoxEditorVB/CustomDropDownEditor.vb
Imports Telerik.WinControls.UI

Public Class CustomDropDownEditor
    Inherits RadDropDownListEditor


    Public Overrides Function EndEdit() As Boolean
        Dim cellElement As GridComboBoxCellElement = TryCast(Me.OwnerElement, GridComboBoxCellElement)
        Dim grid As RadGridView = cellElement.GridControl
        Dim f As RadForm1 = CType(grid.FindForm(), RadForm1)
        Dim dt As NwindDataSet.CategoriesDataTable = f.DataSet.Categories

        For i As Integer = 0 To dt.Rows.Count - 1

            If dt.Rows(i)("CategoryName").ToString() = (CType(Me.EditorElement, RadDropDownListEditorElement)).Text Then
                Return MyBase.EndEdit()
            End If
        Next

        Dim newCategoriesRow As NwindDataSet.CategoriesRow = dt.NewCategoriesRow()
        newCategoriesRow.CategoryName = (CType(Me.EditorElement, RadDropDownListEditorElement)).Text

        f.DataSet.Categories.Rows.Add(newCategoriesRow)
        f.CategoriesTA.Update(f.DataSet.Categories)
        cellElement.Tag = newCategoriesRow.CategoryID
        Return MyBase.EndEdit()
    End Function
End Class

File: GridView/AllowEndUsersAddItemsComboBoxEditor/AllowEndUsersAddItemsComboBoxEditor/CustomDropDownEditor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Telerik.WinControls.UI;

namespace AllowEndUsersAddItemsComboBoxEditor
{
public class CustomDropDownEditor : RadDropDownListEditor
{
       
    public override bool EndEdit()
    {
        GridComboBoxCellElement cellElement = this.OwnerElement as GridComboBoxCellElement;
        RadGridView grid = cellElement.GridControl;
        RadForm1 f = (RadForm1)grid.FindForm();
        // Checking if the typed value exists in the datasource of the column.
        NwindDataSet.CategoriesDataTable dt = f.DataSet.Categories;
        for (int i = 0; i < dt.Rows.Count; i++)
        {
            if (dt.Rows[i]["CategoryName"].ToString() == ((RadDropDownListEditorElement)this.EditorElement).Text)
            {
                return base.EndEdit();
            }
        }
        // An example of what we can do when we enter the custom text.
        // In this case we are adding a new data row in the underlying datasource of 
        // the combobox column and then in the CellEndEdit we are setting
        // the ID value of the newly created row to RadGridView.
        NwindDataSet.CategoriesRow newCategoriesRow = dt.NewCategoriesRow();
        newCategoriesRow.CategoryName = ((RadDropDownListEditorElement)this.EditorElement).Text;
      
        f.DataSet.Categories.Rows.Add(newCategoriesRow);
        // Updating the database. You can do it here at another place
        // you find suitable for this purpose, for example, on FormClosing.
        f.CategoriesTA.Update(f.DataSet.Categories);
        cellElement.Tag = newCategoriesRow.CategoryID;
        return base.EndEdit();
    }
}
}
